iwyu_output.cc

```diff
@@ -1426,6 +1426,71 @@ void CalculateIwyuForFullUse(OneUse* use,
   }
 }
 
+// This function removes all include lines `a` that are desired but not present but
+// transitively included by some include line `b` that is desired and present and
+// adds `a`'s symbol counts to `b`.
+vector<OneIncludeOrForwardDeclareLine> PruneTransitivelyPresent(
+    const vector<OneIncludeOrForwardDeclareLine>& lines,
+    const IwyuPreprocessorInfo* preprocessor_info) {
+  // In the first step, we copy all lines that are an include line, not present,
+  // desired, and whose include_file is transitively included by another line
+  // which is a desired include line into a range `indirect` and the remaining
+  // ones into another range `direct`. This is because elements of range
+  // `indirect` are then precisely those whose symbols are already exposed
+  // through a different header that is to be included anyway, so they can be
+  // omitted.
+  const auto is_transitively_present =
+      [&lines,
+       preprocessor_info](const OneIncludeOrForwardDeclareLine& target) {
+        if (!target.IsIncludeLine() || target.is_present() ||
+            !target.is_desired()) {
+          return false;
+        }
+        // Check if target is transitively included by some relevant source.
+        // This must be proper, equality does not suffice.
+        for (const OneIncludeOrForwardDeclareLine& source : lines) {
+          if (source.included_file() == target.included_file() ||
+              !source.is_desired() || !source.IsIncludeLine()) {
+            continue;
+          }
+
+          if (preprocessor_info->FileTransitivelyIncludes(
+                  source.included_file(), target.included_file())) {
+            return true;
+          }
+        }
+        return false;
+      };
+
+  vector<OneIncludeOrForwardDeclareLine> indirect, direct;
+  indirect.reserve(lines.size());
+  direct.reserve(lines.size());
+  std::partition_copy(lines.cbegin(), lines.cend(), std::back_inserter(indirect),
+                      std::back_inserter(direct), is_transitively_present);
+  // Now add the symbol uses from elements of range indirect to the
+  // corresponding source elements of range direct so that these sources are now
+  // listed as including the corresponding symbols because they do transitively.
+  for (const OneIncludeOrForwardDeclareLine& target : indirect) {
+    const auto source =
+        std::find_if(direct.begin(), direct.end(),
```

Comment:
Hmm. It looks a little off that you need to search for this again -- you've already worked every include line against every other once.

It looks like you already have an O(n^2) algorithm: `partition_copy` traverses all lines, and the lambda you're using to partition traverses all lines. And now this traverses all lines invoking `FileTransitivelyIncludes` again. I wonder if you could flatten this so it only needs to do the N*N loop once?

Reply:
Also think so. If the lambda returned a pointer to the file that transitively includes the target or null, instead of true or false, then a single pass through the lines should be enough. No extra vectors or partition/find_if would be needed either.

The lambda could then be called `transitively_included_by` or smth along those lines.

Reply:
Yes good catch, that's really unnecessary!

Reply:
OK, now that I've looked at it a bit, I'm not sure what the best venue would be.  Returning a pointer isn't really that good of an option because it would still require two iterations over the range: one to add the `target`s' symbols to the `source`s, and one with a `std::remove_if` or so to eliminate these `target`s then.  I don't see how to do this in one go.

One option would be to copy everything into a `vector` of `pair`s of `OneIncludeOrForwardDeclarLine`s and pointers to such, then fill the pointers, then add the symbols, then call `remove_if` on all whose pointer is a `nullptr`, and then `transform` the range back into one where the pointers are removed.  Not sure if I like that option better, also not sure if it's actually faster or slower since it involves more allocations and copying, which might be worse than doing some work twice.

Thoughts?

Reply:
I need to sit down and think about this for a while. It must be possible to make this better than O(N*N) + O(N) (where `FileTransitivelyIncludes` is also non-trivial)

Reply:
I thought briefly about this and I believe it would actually need two passes through the lines to prevent aliasing issues in its current form: one to save the <source,target>pairs and one more to process the vector, and add the symbols from each target to the corresponding source. For the first pass, only the sources that found a valid target would need to be stored (skipping the ones that return null)

That said, I can't actually try anything right now, so please take everything with a grain of salt :)